config/internal/configsource/manager.go

```diff
@@ -457,21 +491,54 @@ func (m *Manager) expandString(ctx context.Context, s string) (interface{}, erro
 	return string(buf) + s[i:], nil
 }
 
-func (m *Manager) retrieveConfigSourceData(ctx context.Context, name, cfgSrcInvoke string) (interface{}, error) {
-	cfgSrc, ok := m.configSources[name]
+// retrieveConfigSourceData retrieves data from the specified config source and injects them into
+// the configuration. The Manager tracks sessions and watcher objects as needed.
+func (m *Manager) retrieveConfigSourceData(ctx context.Context, cfgSrcName, cfgSrcInvocation string) (interface{}, error) {
+	cfgSrc, ok := m.configSources[cfgSrcName]
 	if !ok {
-		return nil, newErrUnknownConfigSource(name)
+		return nil, newErrUnknownConfigSource(cfgSrcName)
 	}
 
-	// Expand any env vars on the selector and parameters. Nested config source usage
-	// is not supported.
-	cfgSrcInvoke = expandEnvVars(cfgSrcInvoke)
-	retrieved, err := m.expandConfigSource(ctx, cfgSrc, cfgSrcInvoke)
+	cfgSrcName, selector, paramsParser, err := parseCfgSrcInvocation(cfgSrcInvocation)
 	if err != nil {
 		return nil, err
 	}
 
-	return retrieved, nil
+	session, ok := m.sessions[cfgSrcName]
+	if !ok {
+		session, err = cfgSrc.NewSession(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create session for config source %q: %w", cfgSrcName, err)
+		}
+		m.sessions[cfgSrcName] = session
+	}
+
+	// Recursively expand the selector.
```

Comment:
Why is this needed? Can selector itself use a nested config source?

Reply:
Since the capability to expand on the parameters was added it seems reasonable to also add the same capability for the selector, example: `include` config source in which the file path is the selector, that now can be an environment variable. IIRC this is one of the cases that AWS wanted to be sure that it worked.